### Visit the Wikipedia hyperlinks graph!

In [1]:
#we create two lists edges and n, one for edges and the last to save al nodes that we find in edges
edges = []
n =[]
#we import data that we need and we fill them 
with open("wiki-topcats-reduced.txt\wiki-topcats-reduced.txt") as mytxt:
    for line in mytxt:
        i,j= line.split()
        n.extend((int(i),int(j)))
        edges.append([int(i),int(j)])
#using set we eliminate duplicates from n and we have a list with the nodes in the graph
nodes= set(n)       

Since edges are directed, the graph is directed

In [2]:
print(len(nodes))

461193


As we can see above the total number of nodes is 461193

In [3]:
print(len(edges))

2645247


Instead 2645247 is the total number of edges

In [4]:
#here we create a dictionary with nodes as keys and degree as attribute, at the begging we set value = for each node
dic_nodes = {key:0 for key in nodes}

#We calculate nodes degree here,considering our list of edges
for e in edges:
    dic_nodes[e[0]]= dic_nodes[e[0]]+1
    dic_nodes[e[1]]= dic_nodes[e[1]]+1

#we calulate the sum of degrees of all nodes
d_sum =0
for k in dic_nodes.keys():
    d_sum += dic_nodes[k]
#finally we have the average node deggre
avg_d = d_sum/len(nodes)
#printing it
print(round(avg_d,2))

11.47


Finally we can say that the average node degree is 11.47, considering the number of nodes this very sparse graph since the maximum number of edges is 461193*461192 = 212698522056